ipn/ipnauth/win.go

```diff
@@ -0,0 +1,205 @@
+			profileUserRights |= Connect | Disconnect | ReadPrefs | ChangeExitNode
+		}
+		if res := checker.grant(profileUserRights); res.HasResult() {
+			return res
+		}
+	}
+
+	if !wi.env.IsServer && !isProfileEnforced(profile, prefs) {
+		// Allow any user to disconnect from non-enforced Tailnets on non-Windows Server devices.
+		// TODO(nickkhyl): automatically disconnect from the current Tailnet
+		// when a different user logs in or unlocks their Windows session,
+		// unless the unattended mode is enabled. But in the meantime, we should allow users
+		// to disconnect themselves.
+		if res := checker.grant(Disconnect); res.HasResult() {
+			return res
+		}
+	}
+
+	if isAdmin, _ := wi.tok.IsAdministrator(); isAdmin {
+		// Allow local admins to disconnect from any tailnet.
+		localAdminRights := Disconnect
+		if wi.env.IsServer || wi.env.IsManaged || wi.tok.IsElevated() {
+			// Allow elevated admins, as well as all admins on managed and server devices,
+			// unrestricted access to all local profiles, except for reading private keys.
+			localAdminRights |= UnrestrictedProfileAccess & ^ReadPrivateKeys
+		}
+		if res := checker.grant(localAdminRights); res.HasResult() {
+			return res
+		}
+	}
+
+	return checker.result()
+}
+
+// Close implements io.Closer by releasing resources associated with the Windows user identity.
+func (wi *WindowsIdentity) Close() error {
+	if wi == nil || wi.tok == nil {
+		return nil
+	}
+	if err := wi.tok.Close(); err != nil {
+		return err
+	}
+	runtime.SetFinalizer(wi, nil)
+	wi.tok = nil
+	return nil
+}
+
+func (wi *WindowsIdentity) checkAdmin() error {
+	isAdmin, err := wi.tok.IsAdministrator()
+	if err != nil {
+		return err
+	}
+	if !isAdmin {
+		return errors.New("the requested operation requires local admin rights")
+	}
+	return nil
+}
+
+func (wi *WindowsIdentity) checkElevatedAdmin() error {
+	if !wi.tok.IsElevated() {
+		return errors.New("the requested operation requires elevation")
+	}
+	return nil
+}
+
+func (wi *WindowsIdentity) isProfileOwner(profile ipn.LoginProfileView) bool {
+	return wi.tok.IsUID(profile.LocalUserID())
+}
+
+func (wi *WindowsIdentity) isProfileUser(profile ipn.LoginProfileView, prefs ipn.PrefsGetter) bool {
+	if profile.LocalUserID() == "" {
+		// Profiles created as LocalSystem (e.g. via MDM) can be used by everyone on the device.
+		return true
+	}
+	if prefs, err := prefs(); err == nil {
+		// Profiles that have unattended mode enabled can be used by everyone on the device.
+		return prefs.ForceDaemon()
+	}
+	return false
+}
+
+func isProfileEnforced(ipn.LoginProfileView, ipn.PrefsGetter) bool {
+	// TODO(nickkhyl): allow to mark profiles as enforced to prevent
+	// regular users from disconnecting.
+	return false
+}
+
+// WindowsEnvironment describes the current Windows environment.
+type WindowsEnvironment struct {
+	IsServer  bool // IsServer indicates whether we're running on a client or server Windows edition.
```

Comment:
either promote this comment to be the preceding line, or make it a fragment like "// whether running on a server edition of Windows"